hugr-model/src/v0/mod.rs

```diff
@@ -0,0 +1,621 @@
+        /// Each of the variants is itself a list of runtime types.
+        variants: TermId,
+    },
+
+    /// The type of functions, given lists of input and output types and an extension set.
+    FuncType {
+        /// The input types of the function, given as a list of runtime types.
+        ///
+        /// `inputs : (list type)`
+        inputs: TermId,
+        /// The output types of the function, given as a list of runtime types.
+        ///
+        /// `outputs : (list type)`
+        outputs: TermId,
+        /// The set of extensions that the function requires to be present in
+        /// order to be called.
+        ///
+        /// `extensions : ext-set`
+        extensions: TermId,
+    },
+
+    /// Control flow.
+    ///
+    /// `(ctrl VALUES) : ctrl` where `VALUES : (list type)`.
+    Control {
+        /// List of values.
+        values: TermId,
+    },
+
+    /// Type of control flow edges.
+    ///
+    /// `ctrl : static`
+    ControlType,
+}
+
+impl<'a> Default for Term<'a> {
+    fn default() -> Self {
+        Self::Wildcard
+    }
+}
+
+/// A parameter to a function or alias.
+///
+/// Parameter names must be unique within a parameter list.
+/// Implicit and explicit parameters share a namespace.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub enum Param<'a> {
+    /// An implicit parameter that should be inferred.
+    Implicit {
+        /// The name of the parameter.
+        name: &'a str,
+        /// The type of the parameter.
+        ///
+        /// This must be a term of type `static`.
+        r#type: TermId,
+    },
+    /// An explicit parameter that should always be provided.
+    Explicit {
+        /// The name of the parameter.
+        name: &'a str,
+        /// The type of the parameter.
+        ///
+        /// This must be a term of type `static`.
+        r#type: TermId,
+    },
+    /// A constraint that should be satisfied by other parameters in a parameter list.
+    Constraint {
+        /// The constraint to be satisfied.
+        ///
+        /// This must be a term of type `constraint`.
+        constraint: TermId,
+    },
+}
+
+/// Errors that can occur when traversing and interpreting the model.
+#[derive(Debug, Clone, Error)]
+pub enum ModelError {
+    /// There is a reference to a node that does not exist.
+    #[error("node not found: {0:?}")]
+    NodeNotFound(NodeId),
+    /// There is a reference to a term that does not exist.
+    #[error("term not found: {0:?}")]
+    TermNotFound(TermId),
+    /// There is a reference to a region that does not exist.
+    #[error("region not found: {0:?}")]
+    RegionNotFound(RegionId),
+    /// There is a local reference that does not resolve.
+    #[error("local variable invalid: {0:?}")]
+    InvalidLocal(String),
+    /// There is a global reference that does not resolve to a node
+    /// that defines a global variable.
+    #[error("global variable invalid: {0:?}")]
+    InvalidGlobal(String),
```

Comment:
should these be taking the relevant `LocalRef` and `GlobalRef` rather than String?

Reply:
In principle yes, and I did that originally. However it would introduce a lifetime parameter to the error, which might not be what we want.